src/Symfony/Component/Clock/HrClock.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Clock;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class HrClock implements TimeInterface
+{
+    private \DateTimeZone $timezone;
+
+    public function __construct(\DateTimeZone|string $timezone = null)
+    {
+        if (\is_string($timezone ??= date_default_timezone_get())) {
+            $this->timezone = new \DateTimeZone($timezone);
+        } else {
+            $this->timezone = $timezone;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function now(): \DateTimeImmutable
+    {
+        $hrtime = hrtime();
+        $now = '@'.$hrtime[0].'.'.substr(str_pad($hrtime[1], 9, 0, \STR_PAD_LEFT), 0, -3);
+
+        return (new \DateTimeImmutable($now, $this->timezone))->setTimezone($this->timezone);
```

Comment:
Setting the timezone in the constructor can be omitted as the object is instantiated using a timestamp - which by definition contains a timezone-declaration so that the second parameter to the constructor is omitted. (see [documentation note](https://www.php.net/manual/en/datetimeimmutable.construct.php#refsect1-datetimeimmutable.construct-parameters))

In addition I'm not sure the extra calls around `hrtime` make the DateTimeImmutable object more precice than calling `return (new DateTimeImmutable())->setTimezone($this->timezone);`

And I'm open for debate whether the `setTimezone()` is necessary at all. But that might be a philosophy topic. Modifying the point of observation of a point in time is IMO not something the Clock should have to take care of.

Reply:
Apart from that I'm getting interesting results from that when testing it in PHP: https://3v4l.org/G4jtQ

And the advantage of the HighResolution timing somehow is eliminated by reducing the nanoseconds to microseconds. As there seems to be no added precision I'd question whether this class is necessary at all...

Reply:
> Setting the timezone in the constructor can be omitted as the object is instantiated using a timestamp - which by definition contains a timezone-declaration so that the second parameter to the constructor is omitted

True, yet it's marginally but still measurably faster to give this argument, that's why I kept it :shrug: 

> In addition I'm not sure the extra calls around hrtime make the DateTimeImmutable object more precice than calling return (new DateTimeImmutable())->setTimezone($this->timezone);

The call to hrtime are not meant to make the timestamp more precise, but to ensure that `now()` and `time*()` methods return monotonic results.

> And I'm open for debate whether the setTimezone() is necessary at all. But that might be a philosophy topic. Modifying the point of observation of a point in time is IMO not something the Clock should have to take care of.

That's a constructor argument so it's implementation specific. I think some ppl will need to set the TZ of all of their DateTimeImmutable objects so I prefer giving control over the TZ to all implementations.

> Apart from that I'm getting interesting results from that when testing it in PHP: [3v4l.org/G4jtQ](https://3v4l.org/G4jtQ)

Yep :)  hrtime() counts "from an arbitrary point in time". Still suitable for relative comparisons.

> And the advantage of the HighResolution timing somehow is eliminated by reducing the nanoseconds to microseconds. As there seems to be no added precision I'd question whether this class is necessary at all...

The other advantage is worth it: monotonicity.
And lets' not forget about the `time*()` methods when needs sub-µs accuracy.